pennylane/transforms/add_noise.py

```diff
@@ -0,0 +1,318 @@
+        TransformProgram(add_noise)
+        >>> qml.transforms.add_noise(circuit, noise_model, level="user").transform_program
+        TransformProgram(cancel_inverses, merge_rotations, undo_swaps, _expand_metric_tensor, add_noise)
+        >>> qml.transforms.add_noise(circuit, noise_model, level="device").transform_program
+        TransformProgram(cancel_inverses, merge_rotations, undo_swaps, _expand_metric_tensor, batch_transform, expand_fn, add_noise)
+
+        Finally, more precise control over exctraction of the transform program at the end of which the transform is to be inserted can be achieved
+        by specifying an integer or slice for indexing. For example:
+
+        >>> qml.transforms.add_noise(circuit, noise_model, level=2).transform_program
+        TransformProgram(cancel_inverses, merge_rotations, add_noise)
+        >>> qml.transforms.add_noise(circuit, noise_model, level=slice(1,3)).transform_program
+        TransformProgram(merge_rotations, undo_swaps, add_noise)
+
+    """
+    if not isinstance(noise_model, qml.NoiseModel):
+        raise ValueError(
+            f"Argument noise_model must be an instance of NoiseModel, got {noise_model}."
+        )
+
+    if level is None or level == "user":  # decompose templates and their adjoints
+
+        def stop_at(obj):
+            if not isinstance(obj, qml.operation.Operator):
+                return True
+            if not obj.has_decomposition:
+                return True
+            return not (hasattr(qml.templates, obj.name) or isinstance(obj, qml.ops.Adjoint))
+
+        error_type = (qml.operation.DecompositionUndefinedError,)
+        decompose = qml.devices.preprocess.decompose
+        [tape], _ = decompose(tape, stopping_condition=stop_at, name="add_noise", error=error_type)
+
+    conditions, noises = [], []
+    metadata = noise_model.metadata
+    for condition, noise in noise_model.model_map.items():
+        conditions.append(lru_cache(maxsize=512)(condition))
+        noises.append(qml.tape.make_qscript(noise))
+
+    new_operations = []
+    for operation in tape.operations:
+        curr_ops = [operation]
+        for condition, noise in zip(conditions, noises):
+            if condition(operation):
+                noise_ops = noise(operation, **metadata).operations
+                if operation in noise_ops and _check_queue_op(operation, noise, metadata):
+                    ops_indx = noise_ops.index(operation)
+                    curr_ops = noise_ops[:ops_indx] + curr_ops + noise_ops[ops_indx + 1 :]
+                else:
+                    curr_ops.extend(noise_ops)
+        new_operations.extend(curr_ops)
+
+    new_tape = type(tape)(new_operations, tape.measurements, shots=tape.shots)
+
+    def null_postprocessing(results):
+        """A postprocesing function returned by a transform that only converts the batch of results
+        into a result for a single ``QuantumTape``.
+        """
+        return results[0]
```

Comment:
All transforms need to return a `postprocessing` function; since there isn't really any post-processing required for this one, we use the standard `null_processing.` But this just reminded me that I could use the one defined in the `devices.preprocess`

Reply:
Yes, I was just wondering why we weren't just returning `results` instead of `results[0]`.